zschema/keys.py

```diff
@@ -30,6 +30,27 @@ def __cmp__(self, other):
             return cmp(int(self.port), int(other.port))
 
 
+# Create a dict with all the keys / values of rhs and lhs, where values
+# from lhs are used if a key is in both.
+def left_merge(lhs, rhs):
+    return dict(v for v in rhs.items() + lhs.items())
+
+
+# Factory for TypeFactorys (e.g. SubRecordType is a TypeFactory for
+# SubRecords).
+class TypeFactoryFactory():
+    def __init__(self, cls, args, kwargs):
```

Comment:
It might be safer to just explicitly not take *args since the merging only really handles kwargs gracefully

Reply:
The `*args` is needed in order to be at all useful for `SubRecord` or `ListOf` (i.e. it is clear what `StringList = ListOf.with_args(String())` does, but it's a lot less clear what `SomeList = ListOf.with_args(doc="List type with unbound object")` is supposed to do).

As it stands now, it doesn't even bother attempting to merge the positional args, under the assumption that  those need to be bound when you create the factory (as in the parentheses above).

Removing it is an option -- we do already have `ListOfType` and `SubRecordType`, and anything else that requires a positional argument *could* be handled similarly -- but, having a unified interface for both cases would be nice.

My preference would be to keep it, though maybe make the behavior more consistent / intuitive.

The most obvious option would be treating the lists like `dict`s with numeric keys, but that allows awful things like `StringList(Integer(), doc="An integer list")`.

The option I'd prefer would be to append them (i.e. `return self.cls(*(self.args + args), ...)`); then you can do things like `GenericListForSomeCategory = ListOf.with_args(category="that category"); my_tagged_string_list = GenericListForSomeCategory(String())`, while `StringList(Integer())` would throw, as it should.